src/ty.md

```diff
@@ -17,110 +179,378 @@ tcx: TyCtxt<'tcx>
 //          arena lifetime
 ```
 
-As you can see, the `TyCtxt` type takes a lifetime parameter.
-During Rust compilation, we allocate most of our memory in
-**arenas**, which are basically pools of memory that get freed all at
-once. When you see a reference with a lifetime like `'tcx`,
-you know that it refers to arena-allocated data (or data that lives as
-long as the arenas, anyhow).
+As you can see, the `TyCtxt` type takes a lifetime parameter. When you see a reference with a
+lifetime like `'tcx`, you know that it refers to arena-allocated data (or data that lives as long as
+the arenas, anyhow).
+
+## Allocating and working with types
+
+To allocate a new type, you can use the various `mk_` methods defined on the `tcx`. These have names
+that correspond mostly to the various kinds of type variants. For example:
+
+```rust,ignore
+let array_ty = tcx.mk_array(elem_ty, len * 2);
+```
+
+These methods all return a `Ty<'tcx>` – note that the lifetime you get back is the lifetime of the
+innermost arena that this `tcx` has access to. In fact, types are always canonicalized and interned
+(so we never allocate exactly the same type twice) and are always allocated in the outermost arena
+where they can be (so, if they do not contain any inference variables or other "temporary" types,
+they will be allocated in the global arena). However, the lifetime `'tcx` is always a safe
+approximation, so that is what you get back.
+
```

Comment:
Would be interested to know examples of when an arena would start and stop - is global=crate level?

Reply:
I believe global means the whole compilation (ie the whole crate graph)

Reply:
@gilescope thanks for noting that. Can you open an issue that ask for that information?. I think would be better if we get this PR in and we continue with follow ups.